panic: runtime error: slice bounds out of range

```
$:~/Downloads/Sia-v1.0.0-darwin-amd64 $ ./siad
Loading...
(1/7) Loading gateway...
(2/7) Loading consensus...
(3/7) Loading transaction pool...
(4/7) Loading wallet...
(5/7) Loading miner...
(6/7) Loading host...
(7/7) Loading renter...
Finished loading in 0.013999521000000001 seconds
panic: runtime error: slice bounds out of range

goroutine 19 [running]:
panic(0x6cb560, 0xc82000a0d0)
	/usr/local/go/src/runtime/panic.go:464 +0x3e6
github.com/NebulousLabs/Sia/modules/gateway.myExternalIP(0x0, 0x0, 0x0, 0x0)
	/home/luke/go/src/github.com/NebulousLabs/Sia/modules/gateway/upnp.go:38 +0x489
github.com/NebulousLabs/Sia/modules/gateway.(*Gateway).threadedLearnHostname(0xc8201881a0)
	/home/luke/go/src/github.com/NebulousLabs/Sia/modules/gateway/upnp.go:62 +0x113
created by github.com/NebulousLabs/Sia/modules/gateway.New
	/home/luke/go/src/github.com/NebulousLabs/Sia/modules/gateway/gateway.go:157 +0xa35
```

--- a/modules/gateway/upnp.go
+++ b/modules/gateway/upnp.go
@@ -7,6 +7,7 @@ import (
 	"net"
 	"net/http"
 	"strconv"
+	"strings"
 	"time"
 
 	"github.com/NebulousLabs/go-upnp"
@@ -29,13 +30,15 @@ func myExternalIP() (string, error) {
 		errResp, _ := ioutil.ReadAll(resp.Body)
 		return "", errors.New(string(errResp))
 	}
-	buf := make([]byte, 64)
-	n, err := resp.Body.Read(buf)
-	if err != nil && err != io.EOF {
+	buf, err := ioutil.ReadAll(io.LimitReader(resp.Body, 64))
+	if err != nil {
 		return "", err
 	}
+	if len(buf) == 0 {
+		return "", errors.New("myexternalip.com returned a 0 length IP address")
+	}
 	// trim newline
-	return string(buf[:n-1]), nil
+	return strings.TrimSpace(string(buf)), nil
 }
 
 // threadedLearnHostname discovers the external IP of the Gateway. Once the IP

--- a/modules/host/upnp.go
+++ b/modules/host/upnp.go
@@ -1,10 +1,13 @@
 package host
 
 import (
+	"errors"
 	"io"
+	"io/ioutil"
 	"net"
 	"net/http"
 	"strconv"
+	"strings"
 	"time"
 
 	"github.com/NebulousLabs/go-upnp"
@@ -23,13 +26,19 @@ func myExternalIP() (string, error) {
 		return "", err
 	}
 	defer resp.Body.Close()
-	buf := make([]byte, 64)
-	n, err := resp.Body.Read(buf)
-	if err != nil && err != io.EOF {
+	if resp.StatusCode != http.StatusOK {
+		errResp, _ := ioutil.ReadAll(resp.Body)
+		return "", errors.New(string(errResp))
+	}
+	buf, err := ioutil.ReadAll(io.LimitReader(resp.Body, 64))
+	if err != nil {
 		return "", err
 	}
+	if len(buf) == 0 {
+		return "", errors.New("myexternalip.com returned a 0 length IP address")
+	}
 	// trim newline
-	return string(buf[:n-1]), nil
+	return strings.TrimSpace(string(buf)), nil
 }
 
 // managedLearnHostname discovers the external IP of the Host. If the host's

